connector/connect/common/src/main/protobuf/spark/connect/commands.proto

```diff
@@ -421,6 +424,39 @@ message StreamingQueryManagerCommandResult {
   }
 }
 
+// The protocol for client-side StreamingQueryListener.
+// This command will only be set when either the first listener is added to the client, or the last
+// listener is removed from the client.
+// The add_listener_bus_listener command will only be set true in the first case.
+// The remove_listener_bus_listener command will only be set true in the second case.
+message StreamingQueryListenerBusCommand {
+  oneof command {
+    bool add_listener_bus_listener = 1;
+    bool remove_listener_bus_listener = 2;
+  }
+}
+
+// The enum used for client side streaming query listener event
+// There is no QueryStartedEvent defined here,
+// it is added as a field in WriteStreamOperationStartResult
+enum StreamingQueryEventType {
+  QUERY_PROGRESS_EVENT = 0;
+  QUERY_TERMINATED_EVENT = 1;
+  QUERY_IDLE_EVENT = 2;
+}
```

Comment:
We need something like `QUERY_EVENT_UNSPECIFIED` as `0`?
https://protobuf.dev/programming-guides/proto3/#enum

Reply:
Also, each variable should have the same prefix as the enum name? I'm not sure if this is a strict rule or not, though. @amaliujia

Reply:
+1 for `QUERY_EVENT_UNSPECIFIED = 0`.

No IIRC we didn't have a strict rule to ask each variable have the same prefix as the enum name.

Reply:
Thanks for the suggestion! let me update